libraries/botbuilder-dialogs/src/dialogContextError.ts

```diff
@@ -34,15 +34,17 @@ export class DialogContextError extends Error {
 
         super(source instanceof Error ? source.message : source);
 
-        this.name = 'DialogContextError';
-
         if (source instanceof Error) {
             this.message = source.message;
             this.stack = source.stack;
         } else {
             this.message = source;
         }
 
+        Object.defineProperty(this, 'name', { enumerable: true, value: 'DialogContextError' });
```

Comment:
I believe we shouldn't be using `defineProperty` due to an issue with IE11 support. Is this an attempt to get the logged `Error` to look nicer? I think @compulim has more context here.

Reply:
I have to mark name/message/stack as `enumerable` to expose the error info because adaptive expressions do not evaluate properties in a JavaScript object if they are `non-enumerable`.
I found this issue because I want to access `turn.dialogEvent.value.name` and `turn.dialogEvent.value.message` in SendActivity in a test case, where the `turn.dialogEvent.value` is a `DialogContextError` object, but the `name` and `message` cannot be evaluated correctly.

Reply:
Can you not do a `turn.dialogEvent instanceof DialogContextError` or even `turn.dialogEvent instanceof Error` to get at those properties?

Reply:
The evaluation of Error objects (or other memory state) is done by adaptive-expressions, but we are not going to check the type in adaptive-expressions, check out this: https://github.com/microsoft/botbuilder-js/blob/57038fcb0984d3df1878e9df3a7066887917ae75/libraries/adaptive-expressions/src/functionUtils.internal.ts#L240
Also, `enumerable` property will also affects `JSON.stringify` behavior, if we don't set this property to `true`, stringify an error would return empty result.

Reply:
[Check this Typescript playground codepad](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD28CiAnF8XQKYA8AuWAdgCYyrqYDeAUNHdAA4CuARiAJbDQpZjHyEQAT2iEwAWywAuaBDwp2hAOYBuarXrABclE2B4MACklQwS6bPmKlAGkasOXHnwHDoLMChk7rASmg09EGyTAxYKMZYpua+ahrB0HgAFuwQAHRiktAAvNAA5AjIaBh5akEAvtSV1FqEctAAZojkGDmiWADucM3FEXkmEGZYeXZ5Hih5sTXa8CBYaSDwSoZNRRRTtRCz84vLAFIAygDyAHJpPsrsDUIrPesbM3MLS4ZHLABWWPppANZYQhBbmsML5YkA), I think if you just make public properties on the Error class it should work properly.

Here's the code directly as well:

```typescript
class FooError extends Error {
    public readonly name: string;

    constructor(message: string, public readonly bar: string) {
        super(message);

        this.name = 'FooError';
    }
}

const fooError = new FooError('message', 'bar');
console.log(fooError);
console.log(JSON.stringify(fooError));
console.log(Object.keys(fooError));
```

Outputs:

```
[LOG]: message 
[LOG]: "{"bar":"bar","name":"FooError"}" 
[LOG]: ["bar", "name"] 
```

Reply:
So maybe we can __re-define__ public properties on `DialogContextError` to make these properties enumerable?

Reply:
Yep, exactly. Give that a shot and see if it works.

Reply:
This should work for name/message properties even without redefining them:
```
class FooError extends Error {
    constructor(source: Error) {
        super();
        this.name = 'FooError';
        this.message = source.message;
        this.stack = source.stack;
    }
}
const fooError = new FooError(new Error('bar'));
console.log(fooError);
console.log(JSON.stringify(fooError));
console.log(Object.keys(fooError));
```
Output:
```
[LOG]: bar 
[LOG]: "{"name":"FooError","message":"bar"}" 
[LOG]: ["name", "message"] 
```
However, stack property is still non-enumerable.

Reply:
Does `FooError` need to derive from the `Error` class? The main reason for that would be to support `instanceof` checks, but it sounds like that may not be strictly necessary in this case.

It's unlikely that we will bundle `botbuilder-dialogs` for IE (the Babel docs specifically cite IE8 as potentially problematic, [see here](https://babeljs.io/docs/en/caveats/)) so if there's no acceptable workaround your can continue to use `Object.defineProperty`.

Reply:
You are the author of `DialogContextError` in https://github.com/microsoft/botbuilder-js/pull/2833/ , we need it to be derived from `Error` because we want to throw it. Is there a way to avoid this?

Reply:
You can technically throw anything, it doesn't need to be an error. I'm wary of changing the class hierarchy now though so let's leave it and just use defineProperty for 'stack'.

Reply:
This isn't applicable to this PR but I wanted to point it out for future feature work.

Because of the way `npm` works regarding dependencies and subDependencies, I'm wary of any cross-library `instanceof` checks. If a bot has `botbuilder-dialogs` as a dependency and subDependency with a `dialogs` version mismatch between the two, `instanceof` checks in the `dialogs`-dependent dependency will fail for classes created from the bot's usage of `dialogs`.

E.g. 
```
bot's node_modules/
- PackageA@1.0.0
- PackageB@1.0.0 // If PackageB does an instanceof on PackageA.ClassA, it will fail
  - PackageA@1.1.0
```

The bot creates ClassA using PackageA@1.0.0 and passes it to PackageB == failed instanceof.

Reply:
This isn't an issue in .NET. But in Python, depending on which package manager you use, you may also run into this problem.

e.g `pip` stores the packages in a flat mechanism but other newer tools allow for dependencies to have different versions of sub-dependencies.

Reply:
It's certainly worth mentioning; however, we don't test our cross-package dependencies with disjoint versions. We could potentially adjust how we express our dependencies (backreferenced peer dependencies or something like that?). For now, I think it's fair to expect that consumers use a single version across all botbuilder packages.

It's similar to how you are expected to use `react` and `react-dom` in lockstep.